Duplicate original id error message

AudreyAltman opened the pull request:
This adds an error/warning message for duplicate `originalId`s.

This also removes the old logic to pretty-print the duplicate `originalId` count, as it is now included in the "Message Summary" part of the pretty printout.

I also confirmed that the duplicate `originalId`s are showing up in the `_LOGS/warnings` csv report. 

This has been tested locally with Oklahoma.

The logic to join the old error messages with the new error messages is a bit convoluted.  The `messages` column in the `DataFrame` of mapped records is an array of `structs`.  I could not find an elegant way to add a new `struct` (i.e. the duplicate original ID message) to an existing array, as the existing spark SQL methods all seem to deal with simpler datatypes.  If anyone has ideas of how to improve on what I landed on, feel free to share.

This addresses `DT-765`

moltude commented on src/main/scala/dpla/ingestion3/executors/MappingExecutor.scala:
```diff
@@ -75,24 +76,50 @@ trait MappingExecutor extends Serializable {
+    // Add a unique identifier (since originalId and dplaId might be duplicated)
+    val resultsWithId: DataFrame = mappingResults
+      .join(duplicateOriginalIds, Seq("originalId"), "outer") // adds column "dupOrigIdMsg"
+      .withColumn("id", monotonically_increasing_id) // adds unique identifier
```
@AudreyAltman -- your problems with adding new messages is what I was banging my head against for a while -- Sorry I can't help there 

@mdellabitta ran against CDL, just about doubled runtime 

**develop**
```
Runtime.............................................................00:06:17.290

Attempted..............................................................1,185,037
Successful.............................................................1,185,036
Failed.........................................................................1


                              Errors and Warnings

Messages
- Errors.......................................................................1
- Warnings.............................................................4,413,880

Records
- Errors.......................................................................1
- Warnings.............................................................1,173,912
```

**duplicate-original-id-error**
```
Runtime.............................................................00:11:00.846

Attempted..............................................................1,185,037
Successful.............................................................1,185,036
Failed.........................................................................1


                              Errors and Warnings

Messages
- Errors.......................................................................1
- Warnings.............................................................4,413,880

Records
- Errors.......................................................................1
- Warnings.............................................................1,173,912
- Duplicate original ID........................................................0
```

AudreyAltman replied:
I think I've landed on a much better solution with simpler logic and shorter runtime.  Let me know what you all think.

AudreyAltman commented:
`oldMessages` column removed, sanity-tested against oklahoma data, rebased to develop

AudreyAltman commented:
I've added two possible solutions (in two separate commits) that address the problem @moltude identified with failing enrichments and reports.

The underlying problem is that order of the columns changed when the new messages were added to the DataFrame.  When the enrichment and reports executors read in the mapped data and converted it to the `DplaMapData` model  via the `ModelConverter`, it expects the columns to be in a specific order.

The first possible solution is to reorder the columns before write, like so: https://github.com/dpla/ingestion3/pull/273/commits/53bd65331adc4d8a20b63b1f8817cc4d02715dc1 . This works fine but I'm not sure I love idea that we need to remember to do this every time we write out `DplaMapData`.

The second possible solution is to change the `ModelConverter` so that it relies on column name rather than column order.  This might be easier to work with in the long run, for us and if we create a tool for other users (not sure the Combine folks might need this?) to work with the `DplaMapData` model. But perhaps there are other considerations I'm not aware of? 
https://github.com/dpla/ingestion3/pull/273/commits/b3fae612f9a801dd1c5270cd37b760b7c6d56e44

I tested both of these locally by running `IngestRemap` on Oklahoma data, comparing console outputs, and spot-checking some reports.  I did not fix the unit tests b/c I wanted to see which solution you all like best before investing time in that.
